runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java

```diff
@@ -65,11 +77,23 @@ public static TestSparkRunner fromOptions(PipelineOptions options) {
     return new TestSparkRunner(sparkOptions);
   }
 
+  /**
+   * Overrides for the test runner.
+   */
+  @SuppressWarnings("unchecked")
   @Override
-  public <OutputT extends POutput, InputT extends PInput>
-      OutputT apply(PTransform<InputT, OutputT> transform, InputT input) {
-    return delegate.apply(transform, input);
-  };
+  public <OutputT extends POutput, InputT extends PInput> OutputT apply(
+          PTransform<InputT, OutputT> transform, InputT input) {
+    // if the pipeline is required to execute as a streaming pipeline,
+    // and the source is an adapted unbounded source (as bounded),
+    // read it as unbounded source via UnboundedReadFromBoundedSource.
+    if (isStreaming && transform instanceof BoundedReadFromUnboundedSource) {
```

Comment:
Yes, definitely. Basically the idea is to replace `PipelineRunner.apply` with a `Pipeline.replace` that takes more-or-less the same parameters and is just as easy to use. @tgroh is actively working on this, doing preparatory refactors to `TransformHierarchy`.

But since the Spark runner already pretty much does this right - doing it in the translation visitor - you have a pretty great hook to just already do it. If it is a pain to do it there, then it is fine to do it here. Porting to the features I am describing should be easy - all runners will need their `apply` to be ported anyhow.